File: Java/src/ShortestWaytoFormString.java
import java.util.ArrayList;
import java.util.Collections;

public class ShortestWaytoFormString {

    public int shortestWay(String source, String target) {
        ArrayList<Integer>[] l = new ArrayList[26];
        for(int i = 0; i < source.length(); i++){
            if(l[source.charAt(i)-'a'] == null) l[source.charAt(i)-'a'] = new ArrayList<>();
            l[source.charAt(i)-'a'].add(i);
        }

        int idx = -1;
        int res = 1;
        for(int i = 0; i < target.length(); i++){
            if(l[target.charAt(i)-'a'] == null) return -1;
            ArrayList<Integer> list = l[target.charAt(i)-'a'];
            int ret = Collections.binarySearch(list, idx+1);
            if(ret >= 0){
                idx = list.get(ret);
            }else{
                int ip = -ret-1;
                if(ip == list.size()) {
                    idx = -1;
                    res++;
                    i--;
                }else{
                    idx = list.get(ip);
                }
            }
        }

        return res;
    }

    public static void main(String[] args) {
        ShortestWaytoFormString test = new ShortestWaytoFormString();
        System.out.println(test.shortestWay("adbsc", "addddddddddddsbc"));
    }

}

File: Python/ShortestWaytoFormString.py
import bisect

class Solution:

    def shortestWay(self, source: str, target: str) -> int:
        map = {}
        for i in range(len(source)):
            if source[i] not in map:
                map[source[i]] = []
            map[source[i]].append(i)
        res = 1
        idx = -1
        i = 0
        while i < len(target):
            if target[i] not in map:
                return -1
            l = map[target[i]]
            ret = bisect.bisect_left(l, idx+1)
            if ret == len(l):
                res += 1
                idx = -1
            else:
                idx = l[ret]
                i += 1
        return res

test = Solution()
print(test.shortestWay("adbsc", "addddddddddddsbc"))

